### DISCUSSION TITLE
Installing dependencies on the fly via the source code might not be best practice

### DISCUSSION THREAD

**Discussion by henriksod** [mikel-brostrom/yolo_tracking#985] [General]
I noticed that dependencies are automatically installed via [TestRequirements](https://github.com/mikel-brostrom/yolo_tracking/blob/b6556a8f435dde494bad29677d0b0d066cbb5ea5/boxmot/utils/checks.py#L29C17-L29C116). I am not sure this is the best idea.

It compromises the users own package installation without their control which might not be desired.

I can see why something like this is needed though. We want to be able to dynamically set dependencies like `onnxruntime` and `onnxruntime-gpu`, as well as not install dependencies that might not be needed by the user (e.g. `nvidia-tensorrt`).

It seems like there are ongoing discussions to support something like this in setuptools [here](https://discuss.python.org/t/idea-selector-packages/4463).

I would instead propose to give a LOG error message to the user when they want to use a feature like [ReIDDetectMultiBackend](https://github.com/mikel-brostrom/yolo_tracking/blob/b6556a8f435dde494bad29677d0b0d066cbb5ea5/boxmot/deep/reid_multibackend.py#L37C7-L37C29) with a non-pytorch weights file.

Example:

> Running ReIDDetectMultiBackend with the specified ONNX weights
> > "\<path\>.onnx"
>
> requires onnxruntime or onnxruntime-gpu to be installed!
>
> Install with one of the following commands:
> For CPU:
> > pip install onnx onnxruntime
>
> For GPU:
> > pip install onnx onnxruntime-gpu

**Comment by mikel-brostrom** [comment]
The non-crucial packages are installed that way. The experience should be as smooth as possible. Giving an error if it is automatable is not an option :smile:

**Reply by henriksod** [reply]
What if I use my global python env and have packages dependent on specific versions, that were to be auto updated running these? IMO being able to install packages dynamically is a design flaw with python.

At least there should be a flag and/or environment variable to toggle automatic installations. Just a thought :-)

**Reply by henriksod** [reply]
I had removed it in #986, but I will revert it then.

It would be easy to fix if it wasn't for the onnxruntime vs onnxruntime-gpu, but can't we have an environment variable or a program argument to activate auto install? It won't of course be possible in the code under boxmot/...

**Reply by mikel-brostrom** [reply]
The idea of envs is that you use one for a specific purpose. Working with a global env is always a bad idea. But I get what you mean. I think this solves the requirements customization based on environment needs: https://python-poetry.org/. Never used it though but seems like a good option.

**Reply by mikel-brostrom** [reply]
But let's discuss this further :smile: before implementing
